Fix single line text not fully rendered

fabOnReact opened the pull request:
## Summary:
### Please re-state the problem that we are trying to solve in this issue.

ReactTextShadowNode#measure calculates the wrong width for single line Text that has an ellipsis (fixes https://github.com/facebook/react-native/issues/39722).

### What is the root cause of that problem?

React Native sets the wrong width in the ReactTextShadowNode [onMeasure][1] function for single line Text, demonstrated with:

[1]: https://github.com/fabriziobertoglio1987/react-native/blob/52ba803c813f4e44a9b5cdc8b2013b725224f08a/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextShadowNode.java#L171 "onMeasure"

- React Native Minimum Reproducible Example ([RNTesterAppShared.js][2]) 
- Android Minimum Reproducible Example ([CustomTextView.java][3], [branch][4])

[2]: https://github.com/fabriziobertoglio1987/react-native/blob/9259392045d6908468eb61ff1d6ce8d109ec13d1/packages/rn-tester/js/RNTesterAppShared.js "react-native mre"
[3]: https://github.com/fabriziobertoglio1987/text-input-cursor-flickering-android/blob/2b727d5ee0b94a1b49f7f3e10e0cd8695f4c67bf/app/src/main/java/com/example/javaexample/CustomTextView.java#L30-L50 "CustomTextView.java"
[4]: https://github.com/fabriziobertoglio1987/text-input-cursor-flickering-android/tree/title-not-full-width "branch android mre"

The android minimum reproducible example:
1) Reproduces the react-native issue using the same implementation of ReactTextShadowNode onMeasure, which calculates the wrong text width with StaticLayout ([screenshot][5])
2) Demonstrates that the default TextView implementation of [onMeasure][6] does not have this issue ([screenshot][7]) (the issue does not reproduce when commenting the implementation of `onMeasure` ). It is not the expected/intended behavior on Android.

[5]: https://user-images.githubusercontent.com/24992535/287243482-a9043b82-311e-4817-b718-cde6ac2f52d7.png "screenshot reproduces the react-native issue"
[6]: https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/widget/TextView.java;l=10852?q=setEllipsizedWidth "screenshot demonstrates that the default TextView implementation of onMeasure does not have this issue"
[7]: https://user-images.githubusercontent.com/24992535/287243426-351b2775-c587-4a86-8c89-8ffa9f475cc0.png

### What changes do you think we should make in order to solve the problem?

Set the correct width for single-line Text in the measure callback.

## Changelog:

[ANDROID] [FIXED] - Fix single line text not fully rendered
## Test Plan:

- [Text width lower than parent view (no ellipsis)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4956212#gistcomment-4956212)
- [Nested Text with different combination of styles (ellipsisMode head)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4956505#gistcomment-4956505)
- [Nested rtl short Text with inline image (image shows to the right)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4958952#gistcomment-4958952)
- [Very long rtl text (image does not show and there is no ellipsis)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4959506#gistcomment-4959506)
- [Text with nested image is shorter than parent view (the image shows and there is no ellipsis)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4959528#gistcomment-4959528)
- [BoringLayout support for margin and padding with single line text](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964433#gistcomment-4964433)
- [The text wraps correctly without leaving white spaces](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964442#gistcomment-4964442)
- [Text width is larger than parent](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964460#gistcomment-4964460)
- [TextView takes the entire line](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964720#gistcomment-4964720)
- [Image does not fix in TextView](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964726#gistcomment-4964726)
- [Images do not fit in the line (parent applies flex style)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964770#gistcomment-4964770)
- [Text with images wraps on the second line](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964800#gistcomment-4964800)
- [Text includes emoji (not boring text)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4965161#gistcomment-4965161)
- [Veryfing the result of the onLayout callback](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4965202#gistcomment-4965202)

fabriziobertoglio1987 commented:
thanks a lot @dmytrorykun. When you do plan to merge this PR?

fabOnReact commented:
Thanks @dmytrorykun, could we land this PR? Thanks

cipolleschi commented on packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/ReactTextShadowNode.java:
```diff
@@ -198,6 +198,11 @@ private Layout measureSpannedText(Spannable text, float width, YogaMeasureMode w
     Layout layout;
     BoringLayout.Metrics boring = BoringLayout.isBoring(text, textPaint);
     float desiredWidth = boring == null ? Layout.getDesiredWidth(text, textPaint) : Float.NaN;
+    // StaticLayout#getLineWidth does not work with single-line text.
+    boolean overrideTextBreakStrategySingleLine =
+        boring == null
+            ? false
+            : mNumberOfLines == 1 && !mAdjustsFontSizeToFit && boring.width > width;
```
What about
```suggestion
    // StaticLayout#getLineWidth does not work with single-line text.
    boolean overrideTextBreakStrategySingleLine =
        boring != null && mNumberOfLines == 1 && !mAdjustsFontSizeToFit && boring.width > width;
```
and we avoid the ternary?

cipolleschi reviewed (commented):
This _seems_ good to me, but I'm not an expert on Android and I have never worked on that part of the infra.
There might be internal tests that might fail with this change. @NickGerleman  @mdvacca do you have time to look into this? The code change seems small but I know that text is delicate.

fabOnReact commented:
<details><summary>Text length higher than parent view width</summary>
<p>



https://github.com/facebook/react-native/pull/41770

<details><summary>CLICK TO OPEN EXAMPLE</summary>
<p>

```javascript
import * as React from 'react';
import {StyleSheet, Text, TextInput, View} from 'react-native';

export default function App() {
  const email =
    'From vincenzoddragon+five@gmail.com  From vincenzoddrlxagon+five@gmail.com';
  return (
    <>
      <View style={styles.container}>
        <View style={styles.flexBrokenStyle}>
          <Text
            style={styles.parentText}
            numberOfLines={1}
          >
            {email}
          </Text>
        </View>
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 8,
    backgroundColor: 'yellow',
  },
  flexBrokenStyle: {
    flexDirection: 'row',
  },
  parentText: {
    backgroundColor: 'red',
  },
});
```



</p>
</details>
 


| Before | OLD ARCH    | NEW ARCH |
| ----------- | ----------- | ----------- |
| <img src="https://gist.github.com/assets/24992535/339ddc3b-a643-4f1a-8e58-91ae04d6e5e6" width="350" />    | <img src="https://gist.github.com/assets/24992535/6068f675-2a0f-4ec4-bee1-818ba22ee671" width="350" />      | <img src="https://gist.github.com/assets/24992535/6d9dcc15-ede3-48a6-b9ed-e24943efce05" width="350" />       |


</p>
</details>

fabOnReact commented:
<details><summary>Text width lower than parent view (no ellipsis)</summary>
<p>

<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

```javascript
import * as React from 'react';
import {StyleSheet, Text, TextInput, View} from 'react-native';

export default function App() {
  const email =
    'From vincenzoddragon+five@gmail.com';
  return (
    <>
      <View style={styles.container}>
        <View style={styles.flexBrokenStyle}>
          <Text
            style={styles.parentText}
            numberOfLines={1}
          >
            {email}
          </Text>
        </View>
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 8,
    backgroundColor: 'yellow',
  },
  flexBrokenStyle: {
    flexDirection: 'row',
  },
  parentText: {
    backgroundColor: 'red',
  },
});
```


</p>
</details>


<image src="https://gist.github.com/assets/24992535/038eaaff-de52-432c-9543-30faf0c18b8f" width="250" />

</p>
</details>

fabOnReact commented:
<details><summary>Nested Text with different combination of styles (ellipsisMode head)</summary>
<p>

<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

```javascript
import * as React from 'react';
import {StyleSheet, Text, TextInput, View} from 'react-native';

export default function App() {
  const email =
    'From vincenzoddragon+five@gmail.com From vincenzoddragon+five@gmail.com';
  return (
    <>
      <View style={styles.container}>
        <View style={styles.flexBrokenStyle}>
          <Text
            android_hyphenationFrequency="full"
            ellipsizeMode="head"
            textBreakStrategy="highQuality"
            style={styles.parentText}
            numberOfLines={1}>
            <Text style={{fontSize: 50}}>
              From vincenzoddragon+five@gmail.com
            </Text>{' '}
            From vincenzoddragon+five@gmail.com
          </Text>
        </View>
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 8,
    backgroundColor: 'yellow',
  },
  flexBrokenStyle: {
    flexDirection: 'row',
  },
  parentText: {
    backgroundColor: 'red',
  },
});
```

</p>
</details>

<image src="https://gist.github.com/assets/24992535/3428f7be-bcef-4f89-a3e7-eaf145619842" width="250" />

</p>
</details>

fabOnReact commented:
<details><summary>Very long rtl text (image does not show and there is no ellipsis)</summary>
<p>


<details><summary>CLICK TO OPEN TESTS RESULTS</summary>
<p>

```javascript
import * as React from 'react';
import {Button, Image, StyleSheet, Text, TextInput, View} from 'react-native';
const fullImage: ImageSource = {
  uri: 'https://www.facebook.com/ads/pics/successstories.png',
};

const smallImage = {
  uri: 'https://www.facebook.com/favicon.ico',
};

export default function App() {
  const email =
    'مرحبا بالعالم من فابريزيومرحبا بالعالم من فابريزيو مرحبا بالعالم من فابريزيومرحبا بالعالم من فابري زيو';
  let textRef = React.useRef(null);
  const measureCallback = (x, y, w, h, pageX, pageY) => {
    console.warn(
      'x: ' +
        parseInt(x) +
        ' y: ' +
        parseInt(y) +
        ' w: ' +
        parseInt(w) +
        ' h: ' +
        parseInt(h),
    );
  };

  const measureTextView = () => {
    if (textRef) {
      textRef.measure(measureCallback);
    }
  };

  return (
    <>
      <View style={styles.container}>
        <View style={styles.flexBrokenStyle}>
          <Text
            ref={ref => (textRef = ref)}
            writingDirection="rtl"
            style={styles.parentText}
            numberOfLines={1}>
            <Text>{email}</Text>
            <Image source={smallImage} style={{height: 50, width: 50}} />
          </Text>
        </View>
        <Button title="measure" onPress={measureTextView} />
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 8,
    backgroundColor: 'yellow',
  },
  flexBrokenStyle: {
    direction: 'rtl',
    flexDirection: 'row',
  },
  parentText: {
    backgroundColor: 'red',
  },
});
```


</p>
</details>

| Before    | After |
| ----------- | ----------- |
| <img src="https://gist.github.com/assets/24992535/dad57fa1-b2e4-4728-b998-182c648c5bea" width="350" />      | <img src="https://gist.github.com/assets/24992535/2aa7c2fd-67a4-416f-bd05-2459f1ad1899" width="350" />       |

</p>
</details>

fabOnReact commented:
- [Text width lower than parent view (no ellipsis)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4956212#gistcomment-4956212))
- [Nested Text with different combination of styles (ellipsisMode head)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4956505#gistcomment-4956505)
- [Nested rtl short Text with inline image (image shows to the right)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4958952#gistcomment-4958952)
- [Very long rtl text (image does not show and there is no ellipsis)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4959506#gistcomment-4959506)
- [Text with nested image is shorter than parent view (the image shows and there is no ellipsis)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4959528#gistcomment-4959528)
- [BoringLayout support for margin and padding with single line text](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964433#gistcomment-4964433)
- [The text wraps correctly without leaving white spaces](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964442#gistcomment-4964442)
- [Text width is larger than parent](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964460#gistcomment-4964460)
- [TextView takes the entire line](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964720#gistcomment-4964720)
- [Image does not fix in TextView](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964726#gistcomment-4964726)
- [Images do not fit in the line (parent applies flex style)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964770#gistcomment-4964770)
- [Text with images wraps on the second line](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4964800#gistcomment-4964800)
- [Text includes emoji (not boring text)](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4965161#gistcomment-4965161)
- [Veryfing the result of the onLayout callback](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4965202#gistcomment-4965202)

fabOnReact commented:
Thanks for the code review. 
> * The premise that we need to use the simpler `BoringLayout` to get the max width text could take without breaking seems strange. `StaticLayout` has options to not line break (maybe we should be setting that for single line instead?)

I updated the solution to use StaticLayout instead of BoringLayout.
> * This will leave the bug when there is a unicode, RTL, or other complicated character.

The solution supports not boring text, emoji, RTL, etc ([test cases](https://github.com/facebook/react-native/pull/41770#issuecomment-1978289361)).
> * This code is duplicated into Fabric (`TextLayoutManager.java` and now some more variants), and this code does not change anything there. If has same issue, we introduce new Fabric regression to Paper. If it isn't there, it has a different solution probably.

I added the Fabric implementation and included [tests](https://github.com/facebook/react-native/pull/41770#issuecomment-1978289361) of the functionalities on Fabric and Paper.

dmytrorykun commented:
@fabOnReact there is a new comment in the corresponding issue https://github.com/facebook/react-native/issues/39722?fbclid=IwAR0G5MSTKkiCcsfqddRDZ-CkLgCEMKER76cBlJvLGQJIOEZa8wqSqGQ_hRE#issuecomment-1992302836

cipolleschi commented:
@realsoelynn this might be related to the measurement issues you are working on.

johnsimeroth commented:
Is this still being worked on / reviewed? Wondering if I need to use a workaround for all the text in my app or if a fix for this issue is likely to be merged before we ship.

realsoelynn commented:
> @realsoelynn this might be related to the measurement issues you are working on.

Sorry about missing this.

This is NOT related to measurement issue that i was previously working on.

@fabOnReact Please also attach `Output of npx react-native info`. 

I saw this issues on https://github.com/Expensify/App/issues/43068 is closed and mentioned it's no longer reproducible.

Do you mind retesting your test cases https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4956212 with latest React Native version with [New Architecture Enabled](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-apps.md#enable-the-new-architecture-on-android) ?

fabOnReact commented:
@realsoelynn 

>Do you mind retesting your test cases https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4956212 with latest React Native version with [New Architecture Enabled](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-apps.md#enable-the-new-architecture-on-android) ?

I tested again the [test case](https://gist.github.com/fabOnReact/6a58c713d32ce5bb9c164b6906abac12?permalink_comment_id=4951474#gistcomment-4951474) with the latest React Native main branch and new architecture enabled.

<details><summary>CLICK TO OPEN SOURCECODE</summary>
<p>

```javascript
import * as React from 'react';
import {StyleSheet, Text, TextInput, View} from 'react-native';

export default function App() {
  const email =
    'From vincenzoddragon+five@gmail.com  From vincenzoddrlxagon+five@gmail.com';
  return (
    <>
      <View style={styles.container}>
        <View style={styles.flexBrokenStyle}>
          <Text
            style={styles.parentText}
            numberOfLines={1}
          >
            {email}
          </Text>
        </View>
      </View>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    padding: 8,
    backgroundColor: 'yellow',
  },
  flexBrokenStyle: {
    flexDirection: 'row',
  },
  parentText: {
    backgroundColor: 'red',
  },
});
```

</p>
</details>

I added a log statement in TextLayoutManager.java and verified that we are running on the new architecture.


<img width="1920" alt="Screenshot 2024-08-10 at 3 49 34 PM" src="https://github.com/user-attachments/assets/000ae168-5eb8-48ef-8e29-383c852c1345">

The width is calculated with TextLayoutManager (newArchitecture), but the text does not take the full width.

<img width="1920" alt="Screenshot 2024-08-10 at 3 50 49 PM" src="https://github.com/user-attachments/assets/4f74e8d1-1534-4d9e-9aaf-8d4dbcf80d10">

Metro output shows Fabric is enabled. 

<img width="945" alt="Screenshot 2024-08-10 at 4 07 22 PM" src="https://github.com/user-attachments/assets/5fa45a23-f9b4-4d82-81f1-0f38911b2175">


>I saw this issues on https://github.com/Expensify/App/issues/43068 is closed and mentioned it's no longer reproducible.

The PR was not created to fix Issue https://github.com/Expensify/App/issues/43068. The Expensify Issue is https://github.com/Expensify/App/issues/21219.

<img src="https://github.com/user-attachments/assets/fd8d5113-e3d9-4f5b-9dc9-5853b21ec0ae" width="350" />    

<img src="https://github.com/user-attachments/assets/e7835c61-3436-45ba-8a5f-45b60101c0ab" width="350" />      



>Please also attach Output of npx react-native info.

<details><summary>OUTPUT OF NPX REACT-NATIVE INFO</summary>
<p>

```
info Fetching system and libraries information...
System:
  OS: macOS 14.6
  CPU: (10) arm64 Apple M1 Pro
  Memory: 176.52 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.15.1
    path: ~/.nvm/versions/node/v20.15.1/bin/node
  Yarn:
    version: 1.22.21
    path: /opt/homebrew/bin/yarn
  npm:
    version: 10.7.0
    path: ~/.nvm/versions/node/v20.15.1/bin/npm
  Watchman:
    version: 2024.07.15.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.3
    path: /Users/fabriziobertoglio/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK: Not Found
IDEs:
  Android Studio: 2023.3 AI-233.14808.21.2331.11709847
  Xcode:
    version: 15.4/15F31d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.12
    path: /usr/bin/javac
  Ruby:
    version: 2.7.5
    path: /Users/fabriziobertoglio/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.73.4
    wanted: 0.73.4
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: true
iOS:
  hermesEnabled: true
  newArchEnabled: true

➜  expensify git:(main) npx react-native info
info Fetching system and libraries information...
System:
  OS: macOS 14.6
  CPU: (10) arm64 Apple M1 Pro
  Memory: 145.38 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.15.1
    path: ~/.nvm/versions/node/v20.15.1/bin/node
  Yarn:
    version: 1.22.21
    path: /opt/homebrew/bin/yarn
  npm:
    version: 10.7.0
    path: ~/.nvm/versions/node/v20.15.1/bin/npm
  Watchman:
    version: 2024.07.15.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.3
    path: /Users/fabriziobertoglio/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.5
      - iOS 17.5
      - macOS 14.5
      - tvOS 17.5
      - visionOS 1.2
      - watchOS 10.5
  Android SDK: Not Found
IDEs:
  Android Studio: 2023.3 AI-233.14808.21.2331.11709847
  Xcode:
    version: 15.4/15F31d
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.12
    path: /usr/bin/javac
  Ruby:
    version: 2.7.5
    path: /Users/fabriziobertoglio/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.73.4
    wanted: 0.73.4
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: true
iOS:
  hermesEnabled: true
  newArchEnabled: true
```

</p>
</details>

s77rt commented:
The problem is that the [created layout](https://github.com/facebook/react-native/blob/7211119d2325a67067f9dd7296cdcb9f91624c7f/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/text/TextLayoutManager.java#L448-L466) renders the text with line break enabled and what we end up measuring is just the first line which does not include all the text (since the rest wrapped). IOW we are measuring the text as if `numberOfLines=undefined`.

<img width="440" alt="Screenshot 2024-11-03 at 11 21 25 PM" src="https://github.com/user-attachments/assets/69b4810e-99f4-4e42-a6fd-9db713a3fce6">
<img width="440" alt="Screenshot 2024-11-03 at 11 21 34 PM" src="https://github.com/user-attachments/assets/6afc4bf6-fc1d-4ca5-b974-fe8b9b899c1f">

s77rt commented:
The suggested solution causes a regression since all Texts with `numberOfLines={1}` will have take full width

| Before | After |
|:------:|:-----:|
|    <img width="440" alt="Screenshot 2024-11-04 at 12 03 46 AM" src="https://github.com/user-attachments/assets/4ae58de7-50e5-4d49-b13f-fb894326ffd3">    |    <img width="440" alt="Screenshot 2024-11-04 at 12 04 46 AM" src="https://github.com/user-attachments/assets/c7b50c8e-0ea1-4a9c-b1af-63ab693d29a4">   |

s77rt commented:
The easiest solution that come to mind is to compare the layout`s number of lines (`layout.getLineCount()`) with the specified `numberOfLines`, if it's greater, then the text is wider than the view, set `calculatedWidth = width`. However this have the same regression above since we can enter `\n` to increase the number of lines but not necessary exceeding the view's width.

I think what we can do is take the number of lines that contain only text and stop at the first one that is just a line break (so we can differentiate between a line wrap and line break).

<img width="440" alt="Screenshot 2024-11-04 at 12 30 32 AM" src="https://github.com/user-attachments/assets/9a1ce7b0-069e-4a75-9c02-dce65d44b590">

s77rt commented:
An alternate solution is to render the text without line break ([setLineBreakConfig](https://developer.android.com/reference/android/text/StaticLayout.Builder#setLineBreakConfig(android.graphics.text.LineBreakConfig))) but I got the app crashed trying to work with this one.
